[WebProfilerBundle] Added feedback about the current symfony version

wouterj opened the pull request:
## Description

This PR adds some visual and textual information about the Symfony version that is currently used. It'll indicate when the used version has reached eom or eol. I hope this will make people more aware of the fact that they should update (as I've seen quite some people using completely outdated Symfony versions).
## Screenshot

![sf-toolbar-version-info](https://cloud.githubusercontent.com/assets/749025/6099512/da59c99e-affa-11e4-8093-173857901769.png)
## PR Information

| Q | A |
| --- | --- |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes (didn't test though) |
| Fixed tickets | - |
| License | MIT |
| Doc PR | - |

wouterj committed:
Added feedback about the current symfony version

jrobeson commented:
why not show the text right under "Symfony Documentation" if the version of EOM/EOL ?

javiereguiluz commented:
Does this change imply to make one HTTP request to symfony.com for each page load in the `dev` environment (including AJAX page loads)?

stloyd commented:
:-1: for calling external website to get such info, even if that would be only on one page (not like now on every page with profiler).

IMO better would be to store such date in Kernel class, same like [version consts](https://github.com/symfony/symfony/blob/2.3/src/Symfony/Component/HttpKernel/Kernel.php#L63), and then compare those in template...

```php
class Kernel
{
    const EOL= '01-06-2015'; // or int 1433109600
}

merk commented:
It could be implemented as an XHR request and use localStorage to cache the information?

javiereguiluz commented:
I think we should not provide this feature. All the information is available at http://symfony.com/roadmap and developers can subscribe to receive emails about important end-of-life announcements.

However, if you insist in adding this feature, I'd add a `Check version` link in the extended information panel. When the user clicks on that link, an AJX request is made to symfony.com to get the information and display it to the user. For example:

![check_symfony_version](https://cloud.githubusercontent.com/assets/73419/6102914/1e307b52-b03e-11e4-871d-1fdebfb0f89a.png)

jrobeson commented:
@javiereguiluz : if the developers were keeping track, then they probably wouldn't be running such EOL versions. I don't think your solution solves the problem

jrobeson commented:
why not just package up the output of the roadmap json and put it in sensiodistributionbundle? and then just parse that?

fabpot commented:
I think having such a built-in feature is a great idea, but making one additional HTTP request to symfony.com for every page in the dev env of all Symfony apps is probably a bit too much.

@javiereguiluz IIUC, the goal is to automatically warn the developer, adding a link is not going to help at all.

@jrobeson An online version is always more flexible as it can take into account last-minute decisions like extending support for one version (because of any external factors).

iltar commented:
I actually like this feature, I often see people come by and ask questions regarding symfony versions that aren't supported anymore. I think this feature should be implemented, but the information should be retrieved when building the container.

jrobeson commented:
sure an online version is more flexible, but as they get a new version of the bundle, then it will include the updated roadmap. i think we shouldn't let the perfect be the enemy of the good here.

jrobeson commented:
@Iltar: i have a problem with symfony requiring internet access unless i explicitly told it to. like on composer updates

WouterJ commented:
The problem with an AJAX request is that'll be catched by the ajax data collector.

I agree that having a request every request isn't very great. We could cache it, so the request will only be executed whenever the cache was cleared?

iltar commented:
@jrobeson in that case, you could turn it off and it would return the gray option: "Unable to retrieve information about the Symfony version.".

Another option could be to show a more specific message: "Unable to retrieve information about the Symfony version, enable to framework.version_check in your configuration".

fabpot commented:
FWIW, I'm strongly against adding a configuration setting for that.

javiereguiluz commented:
I wonder if this is really a "problem" we need to solve:

  * If you care about these things, you are already subscribed to the Symfony notifications, you execute the `security: check` command regularly, you have a continuous integration server that makes a lot of checks, etc.
  * If you don't care about these things, no matter how many times the framework tells you that the version is outdated. You won't update, because you don't care.

merk commented:
@javiereguiluz The issue cant be classified under such black and white terms. It is true most developers who care will be subscribed to notifications, it isnt true that those who are not subscribed do not care and will not update.

There is an opportunity to educate here. A visual notification to indicate the status of symfony has an overall benefit to DX.

Even if it was hard-coded `This version's support expires "roughly" 05/2015` in the Kernel it is hardly a terrible suggestion.

merk commented:
I for one was not aware of a "security: check" command. Not everyone has extensive CI either.

jakzal commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -111,71 +121,101 @@ public function getSymfonyVersion()
+     * @return string the application name
      */
-    public function getAppName()
+    public function getappname()
```
Looks like something went wrong here (and below).

fabpot commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -78,6 +80,14 @@ public function collect(Request $request, Response $response, \Exception $except
             }
         }
+
+        $versionRequest = file_get_contents('http://symfony.com/roadmap.json?version='.$this->data['symfony_version']);
```
What about keeping the result in cache and use the dates from the JSON to "dynamically" determine if the version is out of date or not? That way, we only get the HTTP hit when the version of Symfony changes or when the cache is empty.

WouterJ replied:
Hmm, that sounds like a plan. When the oem date changes, it's probably to make the maintance period longer and not shorter. This means when oem is reached, we can retrieve the data again to see if the dates changed.

stof commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -111,71 +121,101 @@ public function getSymfonyVersion()
+     *
+     * @return string the php version
+     */
+    public function getphpversion()
```
please keep the method name camelcased

javiereguiluz commented:
Just to recap the discussion so far, here are some of my arguments against this feature:

  * Detecting an outdated Symfony version is mostly useful in the production server, where you cannot use this feature.
  * It will slightly hurt application performance because of the added HTTP request.
  * It could severely hit `symfony.com` performance with a flood of requests.
  * There are other (better?) alternatives to get informed when Symfony is out-of date (like the Symfony Roadmap email notifications).
  * No one has ever asked for this feature. In the repository we have hundreds of specific feature requests, but not this one.
  * Anything related to software updates (detecting new versions, notifying about them, installing them, etc.) is one of the most annoying "features" of applications such as operating systems. If you don't believe me, ask around what do people think about the following "features":

![mac_updates_available](https://cloud.githubusercontent.com/assets/73419/6146751/3452d462-b1f0-11e4-8681-f48eff736713.png)

![win_updates_available](https://cloud.githubusercontent.com/assets/73419/6146752/35430b30-b1f0-11e4-9952-b7921a4e2173.jpg)

WouterJ commented:
Thabks for your very opiniated recamp Javier. I think calling it your opinion would be much better.

jakzal commented:
I'm :+1: on this, as long as the requests are cached (on both application and symfony.com).

Comparing this feature to installers is not fair in my opinion. This is just an indicator, it doesn't force you to do anything. It does nothing but sits there (any annoyance coming from that is a good thing).

javiereguiluz commented:
@WouterJ you are right, this is my opinion. That's why the first phrase of my last comment said *"... here are some of my arguments against this feature:"* 

I have yet another reason to not implement this feature:

  * **Privacy**: we are making unsolicited requests to a third-party host (`symfony.com`) without asking permission to the user. This could reveal sensitive information such as the IP address (and therefore, location) of the user.

DavidBadura commented:
what about a console command like `check:version`? everyone can decide for themselves when and whether it should be checked.

WouterJ commented:
In that case, I vote for closing this PR. As executing a command is as simple as navigating to http://symfony.com/roadmap The issue is that many users that use older versions do that because they don't check these tools.

fabpot commented:
@WouterJ I do like this PR and the fact that it tells you that you have a problem **without having to do anything**.

WouterJ commented:
@fabpot I know that, I just responded to @DavidBadura (but forgot the mention...).

javiereguiluz commented:
@fabpot but the price to pay may be too high: making unsolicited HTTP requests to a third-party service (symfony.com) without asking the user and without the possibility of disabling it doesn't look like *"the Symfony way of doing things right"*.

stloyd commented:
@javiereguiluz That's why I proposed to add new simple const to the Kernel class that would hold (already know from roadmap) date of EOL for that version. It's IMO simplest way that will not add any overhead to the profiler. In Twig you could compare dates same it's done in that PR, but without having to call external website...

xabbuh commented:
What about running it as a Composer script (enabling it in the Symfony Standard Edition)? This way, HTTP requests would only be executed when there was network activity at all, and the user would be able to easily disable them by modifying the `composer.json` file.

dunglas commented:
:+1: for this @WouterJ idea too. The solution suggested by @fabpot (querying symfony.com in a CacheWarmer) looks like the good compromise between DX and performance.

WouterJ commented:
I've updated this PR:

 * Results are now queried one the first request and then cached
 * Cache is invalidated when either the state changed (e.g. a version has become eom) or the minor version changed (the developer has updated his symfony packages)
 * Instead of having multiple methods for multiple states, there is now just one method: `getSymfonyState`
 * Added a warning when using a dev version (this is retrieved from `Kernel::EXTRA_VERSION`).

Meanwhile, I removed the lettercase changes (which was a mistake).

fabpot commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -242,4 +258,59 @@ public function getName()
+        if (file_exists($versionCachePath)) {
+            $versionInfo = json_decode(file_get_contents($versionCachePath), true);
+        } else {
+            $versionResponse = file_get_contents('http://symfony.com/roadmap.json?version='.$this->data['symfony_version']);
```
IIRC, `file_get_contents` can emits some warnings.

WouterJ replied:
Indeed:

 > An E_WARNING level error is generated if filename cannot be found, maxlength is less than zero, or if seeking to the specified offset in the stream fails.

Should I prefix it with `@`?

fabpot replied:
Yes

stof commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -242,4 +258,59 @@ public function getName()
+            $eom = \DateTime::createFromFormat('m/Y', $versionInfo['eom'])->modify('last day of this month');
+            $eol = \DateTime::createFromFormat('m/Y', $versionInfo['eol'])->modify('last day of this month');
+
+            if ('-' === $now->diff($eom)->format('%r')) {
```
``if ($now > $eom)`` ?

WouterJ commented:
Updated the PR to apply the 2 comments, thanks! However, for some reason, Travis decided to not run the builds for my PRs...

fabpot commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -242,4 +258,59 @@ public function getName()
+
+        // get the version state
+        $versionState = 'unknown';
+        if (isset($versionInfo)) {
```
Instead of using isset here, I would prefer that we always define the `$versionInfo` variable and test for `null` here.

WouterJ replied:
fixed

fabpot replied:
hmmm, not yet :) Forgot to push?

WouterJ replied:
I fixed it already locally, but I first need to install the vendors to fix the tests. If that succeeds, I'll push.

(my comment was mainly a note to myself that this "todo" item was done).

stof commented:
@WouterJ it ran. Just be patient. Travis allows only 5 concurrent job per repo owner, so the queue of Symfony builds can take some time when contributions are active

stof commented:
and btw, tests are failing

WouterJ commented:
I've disabled caching in the tests, as it was causing problems. Maybe we should disable requesting the version completely, to not flood symfony.com with requests? /cc @fabpot

WouterJ commented:
Test failures seem unrelated

fabpot commented on src/Symfony/Component/HttpKernel/DataCollector/ConfigDataCollector.php:
```diff
@@ -242,4 +264,63 @@ public function getName()
+        if (file_exists($versionCachePath)) {
+            $versionInfo = json_decode(file_get_contents($versionCachePath), true);
+        } else {
+            $versionResponse = @file_get_contents('http://symfony.com/roadmap.json?version='.preg_replace('/-\w+\d*$/', '', $this->data['symfony_version']));
```
You should pass a version like `x.y`. A `x.y.z` will be anyway redirected to `x.y`. That would reduce the number of requests to symfony.com.

fabpot commented:
@WouterJ No need to disable for the tests, we have some cache on this resource. After fixing my comment, I'm :+1:

WouterJ commented:
Ok, fine by me. I've fixed your comment (and pushed).

WouterJ commented:
Pushed one final safety guard: If the version was wrong for some reason and the error page (http://symfony.com/roadmap.json?version=2.4.x) was returned, it'll also show the "Unable to retrieve information" message.

fabpot commented:
Thank you @WouterJ.

sprain commented:
I'm a bit late to the party, but an option to turn this non-vital feature off per installation would be nice. As seen in the comments above, some users don't like it for privacy reasons or other concerns. Why not give them the option to do so?

shoomyth commented:
This must be disableable

King2500 commented:
@fabpot Can you explain why you are against adding a configuration setting for that?

javiereguiluz commented:
One additional issue with this feature is that it's only available for 2.7. The irony is that 2.7 version is supported until May 2018. This means that your application will make thousands of HTTP requests during the next years to tell you that 2.7 version is still supported :cry:

WouterJ commented:
please, @javiereguiluz, can you read the discussion and code? Your application will make just 1 request and 1 again on 31 May 2018.. Nice hu?!

The rest, I'm waiting on feedback from the core guys here. As @fabpot made it very clear that he doesn't like a config option for this, I'm not putting effort in creating a PR yet.

slider commented:
@fabpot Are requests to http://symfony.com/roadmap.json stored/analyzed in some way on the serverside?

derrabus commented:
I missed thw whole discussion on this feature and just became aware of it because of the bug report I've just filed (#14349).

Please reconsider this feature, for all the reasons @javiereguiluz listed in this discussion. It's a functionality I would never have asked for. I want to be able to work in environments without permanent internet connection and I really don't want libraries to call home.

hhamon commented:
Can't we have the end of life / end of maintainance dates in the composer.json file of Symfony? This way it's updated when running `composer update` command. As Composer will always have to go on the Internet, this may be the right moment to update those info and cache them somewhere (cache directory or composer.lock file).

derrabus commented:
The roadmap of Symfony is pretty static. We already know EOM and EOL dates of the 2.7 branch, so we could simply add this information as constants to the `Symfony\Component\HttpKernel\Kernel` class, right below the other version constants.

If those dates change in the future, we would update them and ship them with the next release. I don't expect the maintenance period to be reduced afterwards, so the worst thing that might happen would be that the guy who is running 2.7.16 although 2.7.42 is the most current one gets a notification that his version is out of support, although the maintenance period has been extended with the 2.7.21 release. But he's running an old version anyway, so I could live with that.

apfelbox commented:
I very much like and support the reasoning from @derrabus and would love to see it implemented. Although I probably would include it in a separate composer package. This would work the same way as @derrabus explained it for the constants in the kernel – as in that the user gets a notification of his old version, regardless whether something has changed in the roadmap or not. So (s)he would issue a composer update which would automatically fetch the new version of the roadmap package (with the new version data).

This library would mainly extract the logic and data from http://symfony.com/roadmap into its own package. 

This package could optionally even include the dynamic (and cached) call, which would just override the locally (inside the package) stored default version data.
